libsnark/zk_proof_systems/plonk/utils.hpp

```diff
@@ -196,6 +196,56 @@ std::vector<std::vector<FieldT>> plonk_gates_matrix_transpose(
     const size_t &nrows,
     const size_t &ncols);
 
+/// deterministically compute values for the constants k1,k2 \in Fr
+/// (section 8.1 [GWC19]) as k1=g^{2s}, k2=nqr, where s is the largest
+/// power of 2 such that 2^s < p and 2^{s+1} > p, p is the prime
+/// modulus of Fr and nqr is a quadratic nonresidue in Fr^*. both s
+/// and nqr are members of the class Fp_model defined in
+/// libff/libff/algebra/fields/fp.hpp .
+template<typename FieldT>
+void plonk_generate_constants_k1_k2(FieldT &k1_result, FieldT &k2_result);
+
+/// generate values for the constants k1,k2 \in Fr (section 8.1
+/// [GWC19]) by trying random values until the following three
+/// conditions are simultaneously satisfied:
+///
+/// 1) k1^n != 1 ensuring that k1 \notin H
+/// 2) k2^n != 1 ensuring that k2 \notin H
+/// 3) (k1 k2^-1)^n != 1 ensuring that k2H \notin k1H (and vice-versa)
+template<typename FieldT>
+void plonk_generate_random_constants_k1_k2(
+    FieldT &k1_result, FieldT &k2_result);
+
+/// check that the constants k1,k2 \in Fr (section 8.1 [GWC19]) are
+/// valid. this is the case if the sets H, k1H, k2H are
+/// non-overlapping, where H = {w, w^1, w^2, ..., 2^n} is a
+/// multiplicative subgroup of Fr of order n with generator w (denoted
+/// as \omega in [GWC19]). w is a primitive n-th root of unity in Fr
+/// and n is of the form 2^s where s is the largest power of 2 such
+/// that 2^s < p and 2^{s+1} > p, where p is the prime modulus of
```

Comment:
Already called the field Fr.
```suggestion
/// that 2^s | r-1 where r is the prime modulus of
```

Reply:
Fixed.